python/tvm/contrib/pyfrontend.py

```diff
@@ -0,0 +1,305 @@
+"""Compiling a subset of Python to HalideIR"""
+import ast
+import inspect
+import types
+import operator
+#import numpy
+from .. import expr as _expr
+from .. import stmt as _stmt
+from .. import make as _make
+from .. import api  as _api
+from .. import ir_pass  as _ir_pass
+from .. import build_module as builder
+
+#pylint: disable=missing-docstring, invalid-name, consider-merging-isinstance, no-else-return
+#pylint: disable=inconsistent-return-statements, eval-used
+
+NOOP = _make.Evaluate(_api.const(0))
+
+class Unrolled(object):
+    def __init__(self, factor=1):
+        pass
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, a, b, c):
+        return self
+
+#TODO: uncomment these lines later to support intermediate buffer
+#def allocate(shape, dtype=None):
+#    return numpy.zeros(shape).tolist()
+
+def py_frontend(func):
+    if 'Unrolled' not in func.__globals__.keys():
+        func.__globals__['Unrolled'] = Unrolled
+    return func
+
+def _list_to_block(lst):
+    lst = [stmt for stmt in lst if not _ir_pass.Equal(stmt, NOOP)]
+    if not lst:
+        return NOOP
+    if len(lst) == 1:
+        return lst[0]
+    body = lst[0]
+    for i in lst[1:]:
+        body = _make.Block(body, i)
+    return body
+
+class PyVariableUsage(ast.NodeVisitor):
+
+    def __init__(self):
+        self.declared_vars = {}
+        self.loops_above = []
+        self.in_loop_decl = False
+
+    def visit_For(self, node):
+        self.in_loop_decl = True
+        self.visit(node.target)
+        self.in_loop_decl = False
+
+        for i in node.body:
+            self.visit(i)
+
+        self.loops_above.pop()
+
+    def visit_Name(self, node):
+        if self.in_loop_decl:
+            self.loops_above.append(node.id)
+            return
+
+        # The loop variable cannot be overwritten when iteration
+        if isinstance(node.ctx, ast.Store):
+            assert node.id not in self.loops_above
+
+        if node.id not in self.declared_vars:
+            if isinstance(node.ctx, ast.Store):
+                self.declared_vars[node.id] = set()
+        else:
+            self.declared_vars[node.id].add(type(node.ctx))
+
+def _find_all_variable_decl(ir):
+    visitor = PyVariableUsage()
+    visitor.visit(ir)
+    return visitor.declared_vars
+
+class PyAST2HalideIR(ast.NodeVisitor):
+
+    _binop_maker = {
+        ast.Add   : _make.Add,
+        ast.Sub   : _make.Sub,
+        ast.Mult  : _make.Mul,
+        ast.Div   : _make.Div,
+        ast.Mod   : _make.Mod,
+        ast.BitOr : operator.or_,
```

Comment:
A question here,  we have all operations above in halide, why we have two different ir building here?

Reply:
This is only for parsing and building IR node.